packages/core/src/2d/text/TextUtils.ts

```diff
@@ -0,0 +1,309 @@
+import { OverflowMode } from "../enums/TextOverflow";
+
+/**
+ * TextContext.
+ */
+export interface TextContext {
+  canvas: HTMLCanvasElement;
+  context: CanvasRenderingContext2D;
+}
+
+/**
+ * TextMetrics.
+ */
+export interface TextMetrics {
+  width: number;
+  height: number;
+  lines: Array<string>;
+  lineWidths: Array<number>;
+  maxLineWidth: number;
+  lineHeight: number;
+  fontSize: number;
+}
+
+/**
+ * TextUtils includes some helper function for text.
+ */
+export class TextUtils {
+  /** @internal */
+  static _genericFontFamilies: Array<string> = [
+    "serif",
+    "sans-serif",
+    "monospace",
+    "cursive",
+    "fantasy",
+    "system-ui",
+    "math",
+    "emoji",
+    "fangsong"
+  ];
+  /** These characters are all tall to help calculate the height required for text. */
+  private static _testString: string = "|ÉqÅ";
+  private static _testBaseline: string = "M";
+  private static _heightMultiplier: number = 2;
+  private static _baselineMultiplier: number = 1.4;
+  private static _maxWidth: number = 2048;
+  private static _maxHeight: number = 2048;
+  private static _pixelsPerUnit: number = 128;
+  private static _fontSizeCache: Record<string, number> = {};
+  private static _textContext: TextContext = null;
+
+  /**
+   * The instance function to get an object includes 2d context and canvas.
+   * @returns the TextContext object
+   */
+  public static textContext(): TextContext {
+    if (!TextUtils._textContext) {
+      const canvas = document.createElement("canvas");
```

Comment:
check the benchmarks: https://www.measurethat.net/Benchmarks/Show/12247/2/setfont-measuretext-offscreencanvas-vs-canvas#latest_results_block